src/server/internalServer.cpp

```diff
@@ -678,30 +697,171 @@ InternalServer::search_catalog(const RequestContext& request,
+  } else {
+    return Response::build_404(*this, request, "");
+  }
+}
+
+std::unique_ptr<Response> InternalServer::handle_catalog_v2_root(const RequestContext& request)
+{
+  const std::string root_url = normalizeRootUrl(m_root);
+  return ContentResponse::build(
+             *this,
+             RESOURCE::catalog_v2_root_xml,
+             kainjow::mustache::object{
+               {"date", gen_date_str()},
+               {"endpoint_root", root_url + "/catalog/v2"},
+               {"feed_id", gen_uuid(m_library_id)},
+               {"all_entries_feed_id", gen_uuid(m_library_id + "/entries")},
+               {"category_list_feed_id", gen_uuid(m_library_id + "/categories")}
+             },
+             "application/atom+xml;profile=opds-catalog;kind=navigation"
+  );
+}
+
+std::unique_ptr<Response> InternalServer::handle_catalog_v2_entries(const RequestContext& request)
+{
+  const std::string root_url = normalizeRootUrl(m_root);
+
+  const auto now = gen_date_str();
+  kainjow::mustache::list bookData;
+  const auto filter = get_search_filter(request);
+  const auto allMatchingEntries = mp_library->filter(filter);
+  const size_t count = request.get_optional_param("count", 10UL);
+  const size_t start = request.get_optional_param("start", 0UL);
+  for ( const auto& bookId : subrange(allMatchingEntries, start, count) ) {
+    const Book& book = mp_library->getBookById(bookId);
+    const MustacheData bookUrl = book.getUrl().empty()
+                               ? MustacheData(false)
+                               : MustacheData(book.getUrl());
+    bookData.push_back(kainjow::mustache::object{
+      {"id", "urn:uuid:"+book.getId()},
+      {"name", book.getName()},
+      {"title", book.getTitle()},
+      {"description", book.getDescription()},
+      {"language", book.getLanguage()},
+      {"content_id",  book.getHumanReadableIdFromPath()},
+      {"updated", book.getDate() + "T00:00:00Z"},
+      {"category", book.getCategory()},
+      {"flavour", book.getFlavour()},
+      {"tags", book.getTags()},
+      {"article_count", to_string(book.getArticleCount())},
+      {"media_count", to_string(book.getMediaCount())},
+      {"author_name", book.getCreator()},
+      {"publisher_name", book.getPublisher()},
+      {"url", bookUrl},
+      {"size", to_string(book.getSize())},
+    });
+  }
+
+  const auto query = request.get_query().empty()
+                   ? MustacheData(false)
+                   : MustacheData(request.get_query());
+
+  return ContentResponse::build(
+             *this,
+             RESOURCE::catalog_v2_entries_xml,
+             kainjow::mustache::object{
+               {"date", now},
+               {"endpoint_root", root_url + "/catalog/v2"},
+               {"feed_id", gen_uuid(m_library_id + "/entries?"+request.get_query())},
+               {"filter", query},
+               {"books", bookData }
+             },
+             "application/atom+xml;profile=opds-catalog;kind=acquisition"
+  );
```

Comment:
Even if this is not mandatory by opds spec, it would be better to also
add openSearch element (https://github.com/dewitt/opensearch/blob/master/opensearch-1-1-draft-6.md#elements)
(As OPDSDumper does)

---

If you need to search for opensearch spec, it seems that now opensearch.org is use for the opensearch software (:facepalm: :angry: :cry:). So all links you can find on the web are wrong, especially the ones in opds specs. The opensearch spec is now at https://github.com/dewitt/opensearch

Reply:
Done